.github/workflows/image-builder.yml

```diff
@@ -0,0 +1,181 @@
+        required: false
+        type: string
+        default: '[]'
+
+jobs:
+  build-base-images:
+    strategy:
+      max-parallel: 5
+      matrix:
+        image: ${{ fromJSON(inputs.base-images) }}
+        os: [ubuntu-24.04, ubuntu-24.04-arm]
+        exclude: ${{ fromJSON(inputs.exclude) }}
+    runs-on: ${{ matrix.os }}
+    env:
+      CAN_PUSH_IMAGE: ${{ secrets.DOCKERHUB_USERNAME != '' && github.ref_name == github.event.repository.default_branch }}
+    steps:
+      # Checkout the repository with full depth so we can check if the workspace was modified
+      - uses: actions/checkout@v4
+        with:
+          fetch-depth: 0
+
+      - name: Check whether ${{ inputs.folder }}/${{ matrix.image }} was modified
+        id: check-modified
+        run: |
+          scripts/check_path_modified.sh ${{ inputs.folder }}/${{ matrix.image }} ${{ inputs.folder }}_${{ matrix.image }}_modified
+
+      - name: Set up Docker Buildx
+        # Currently, we need this 'if' on each step. https://github.com/orgs/community/discussions/82744
+        if: ${{ steps.check-modified.outputs.was-modified == 'true' }}
+        uses: docker/setup-buildx-action@v3
+
+      - name: Login to DockerHub
+        if: ${{ steps.check-modified.outputs.was-modified == 'true' && env.CAN_PUSH_IMAGE == 'true' }}
+        uses: docker/login-action@v3
+        with:
+          username: ${{ secrets.DOCKERHUB_USERNAME }}
+          password: ${{ secrets.DOCKERHUB_TOKEN }}
+
+      - name: Set docker tags and labels
+        id: meta
+        if: ${{ steps.check-modified.outputs.was-modified == 'true' }}
+        uses: docker/metadata-action@v5
+        with:
+          context: git
+          images: |
+            ${{ inputs.image-prefix }}-${{ matrix.image }}
+          # On the default branch, we'll also tag the image as `latest`
+          tags: |
+            type=raw,value=latest,enable={{is_default_branch}}
+            type=sha,prefix=,format=long
+
+      - name: Build ${{ (env.CAN_PUSH_IMAGE == 'true' && 'and push') || '' }}
+        id: build
+        if: ${{ steps.check-modified.outputs.was-modified == 'true' }}
+        uses: docker/build-push-action@v6
+        with:
+          context: ${{ inputs.folder }}/${{ matrix.image }}
+          labels: ${{ steps.meta.outputs.labels }}
+          tags: ${{ steps.meta.outputs.tags }}
+          push: ${{ env.CAN_PUSH_IMAGE == 'true' }}
+          outputs: |
+            type=docker,dest=${{ runner.temp }}/${{ matrix.image }}-${{ matrix.os }}.tar
+
+      # We will upload the artifact so we can use it in the next job
+      - name: Upload artifact
+        if: ${{ steps.check-modified.outputs.was-modified == 'true' }}
+        uses: actions/upload-artifact@v4
+        with:
+          name: ${{ matrix.image }}-${{ matrix.os }}-tarball
+          path: ${{ runner.temp }}/${{ matrix.image }}-${{ matrix.os }}.tar
+          if-no-files-found: error
+          retention-days: 1
```

Comment:
This is the minimum retention. https://github.com/actions/upload-artifact?tab=readme-ov-file#retention-period